Fix finding document entries with a bad pacer_doc_id

ttys0dev opened the pull request:
I think we need a fallback here in case we have entries with bad `pacer_doc_id`'s.

mlissner commented:
Thanks! One day, we'll catch all these variations, and retire to a beach, I suppose. Until then, Alberto, can you please review? Seems like we'll want a test for this as well?

albertisfu reviewed (changes requested):
Thanks @ttys0dev  I've added a couple of related tests. They're currently failing due to two issues that may require to be addressed:

1. When removing the invalid `pacer_doc_id` from the fallback lookup, the `RECAPDocument` is matched but retains its previous invalid `pacer_doc_id`. Should we update the incorrect `pacer_doc_id` with the new one? Currently, the `pacer_doc_id` is assigned here:
https://github.com/freelawproject/courtlistener/blob/3060ac0b983894e630075917a8ad37da8f126a31/cl/recap/mergers.py#L965

   While prioritizing the current `pacer_doc_id` is generally expected behavior, in this case we'd need to update it from the `docket_entry` if it's not blank.

2. The fallback lookup without `pacer_doc_id` can raise a `MultipleObjectsReturned` error if there is more than one main PACER Document in the docket entry. Should we apply the same logic to select the latest RECAPDocument with PDF and remove duplicates?
https://github.com/freelawproject/courtlistener/blob/3060ac0b983894e630075917a8ad37da8f126a31/cl/recap/mergers.py#L948
   We could create a method for this logic and use it in both try/except blocks.

ttys0dev commented:
> 1. Should we update the incorrect `pacer_doc_id` with the new one?

I think so, added handling for this.

> We could create a method for this logic and use it in both try/except blocks.

Yeah, cleaned this up a bit by making a function for removing duplicates.

albertisfu commented on cl/recap/mergers.py:
```diff
@@ -822,6 +822,19 @@ async def get_or_make_docket_entry(
 
 
+async def clean_duplicate_documents(params) -> RECAPDocument:
+    duplicate_rd_queryset = RECAPDocument.objects.filter(**params)
+    rd_with_pdf_queryset = duplicate_rd_queryset.filter(
+        is_available=True
+    ).exclude(filepath_local="")
+    if await rd_with_pdf_queryset.aexists():
+        rd = await rd_with_pdf_queryset.alatest("date_created")
+    else:
+        rd = await duplicate_rd_queryset.alatest("date_created")
+    await duplicate_rd_queryset.exclude(pk=rd.pk).adelete()
+    return rd
```
I noticed that we still have a related duplicated code fragment in `clean_duplicate_attachment_entries`
https://github.com/freelawproject/courtlistener/blob/0c979de9f364dab7224adb677ecc9c85da439bef/cl/recap/mergers.py#L1624

My suggestion is to split the logic into two methods, keeping this one as something like:

```suggestion
    duplicate_rd_queryset = RECAPDocument.objects.filter(**params)
    return await keep_latest_rd_document(duplicate_rd_queryset)
```

And create a new method e.g: `keep_latest_rd_document` method with the remaining logic:

```
async def keep_latest_rd_document(queryset:QuerySet) -> RECAPDocument:
    rd_with_pdf_queryset = queryset.filter(
            is_available=True
        ).exclude(filepath_local="")
    if await rd_with_pdf_queryset.aexists():
        rd = await rd_with_pdf_queryset.alatest("date_created")
    else:
        rd = await queryset.alatest("date_created")
    await queryset.exclude(pk=rd.pk).adelete()
    return rd
```

Therefore, we can remove the duplicated code in `clean_duplicate_attachment_entries` by passing a different queryset.
```
duplicate_rd_queryset = rds.filter(pacer_doc_id=dupe.pacer_doc_id)
keep_rd = await keep_latest_rd_document(duplicate_rd_queryset)
```

ttys0dev replied:
applied as suggested

albertisfu commented on cl/recap/mergers.py:
```diff
@@ -822,6 +822,19 @@ async def get_or_make_docket_entry(
     return de, de_created
 
 
+async def clean_duplicate_documents(params) -> RECAPDocument:
```
We just need to add the types and docstrings, please.

```suggestion
async def clean_duplicate_documents(params:dict[str, Any]) -> RECAPDocument:
    """ Removes duplicate RECAPDocuments, keeping the most recent with PDF if
    available or otherwise the most recent overall.

    :param params: Query parameters to filter the RECAPDocuments.
    :return: The matched RECAPDocument after cleaning.
    """
```

albertisfu reviewed (commented):
@ttys0dev thank you! Changes look good. 

Just a couple of remaining comments, please.

albertisfu commented on cl/recap/mergers.py:
```diff
@@ -934,35 +963,39 @@ async def add_docket_entries(
+                        is_available=False,
+                        **params,
+                    )
+                    rds_created.append(rd)
```
I moved this line `rds_created.append(rd)` to apply only in cases where the RD is created. Previous approach was also applying it to existing RDs.

albertisfu reviewed (approved):
Thanks @ttys0dev for the changes. I've made a small tweak to a docstring and moved the line `rds_created.append(rd)` as explained in a comment.

@mlissner this looks good to be merged.
